include/internal/list.h

```diff
@@ -41,30 +62,42 @@
+        OSSL_LIST_CHK(list->alpha == NULL                                   \
+                      || list->alpha->ossl_list_ ## name.list == list)      \
+        return list->alpha;                                                 \
     }                                                                       \
     static ossl_unused ossl_inline type *                                   \
     ossl_list_##name##_tail(const OSSL_LIST(name) *list)                    \
     {                                                                       \
-        return list->tail;                                                  \
+        OSSL_LIST_CHK(list->omega == NULL                                   \
+                      || list->omega->ossl_list_ ## name.list == list)      \
+        return list->omega;                                                 \
     }                                                                       \
     static ossl_unused ossl_inline type *                                   \
     ossl_list_##name##_next(const type *elem)                               \
     {                                                                       \
+        OSSL_LIST_CHK(elem->ossl_list_ ## name.next == NULL                 \
+                      || elem->ossl_list_ ## name.next                      \
+                             ->ossl_list_ ## name.prev == elem)             \
         return elem->ossl_list_ ## name.next;                               \
     }                                                                       \
     static ossl_unused ossl_inline type *                                   \
     ossl_list_##name##_prev(const type *elem)                               \
     {                                                                       \
+        OSSL_LIST_CHK(elem->ossl_list_ ## name.prev == NULL                 \
+                      || elem->ossl_list_ ## name.prev                      \
+                             ->ossl_list_ ## name.next == elem)             \
         return elem->ossl_list_ ## name.prev;                               \
     }                                                                       \
     static ossl_unused ossl_inline void                                     \
     ossl_list_##name##_remove(OSSL_LIST(name) *list, type *elem)            \
     {                                                                       \
-        if (list->head == elem)                                             \
-            list->head = elem->ossl_list_ ## name.next;                     \
-        if (list->tail == elem)                                             \
-            list->tail = elem->ossl_list_ ## name.prev;                     \
+        OSSL_LIST_CHK(elem->ossl_list_ ## name.list == list)                \
+        OSSL_LIST_DBG(elem->ossl_list_ ## name.list = NULL)                 \
```

Comment:
Not sure why this is helpful?

Reply:
It's preventing the removal of an element from a list twice (assert on the second).  That seems kind of useful to me.

YMMV.